include/nil/crypto3/math/domains/detail/basic_radix2_domain_aux.hpp

```diff
@@ -47,9 +47,11 @@ namespace nil {
                 void basic_radix2_fft(Range &a, const typename FieldType::value_type &omega) {
                     typedef typename std::iterator_traits<decltype(std::begin(std::declval<Range>()))>::value_type
                         value_type;
-
+                    typedef typename FieldType::value_type field_value_type;
```

Comment:
As I understand we can make operations only if `Range &a` has same `value_type` as `FieldType::value_type`. So whats why we use everywhere  `Range::value_type` and check it in assert with `FieldType::value_type`
But maybe @nkaskov has more information about this...

Reply:
The idea is, that since ffi is a linear, function it can in be used on curve elements too, which support addition as well as multiplication by  scalar field elements.

So in this case field value type and range value type are not the same, and range value type is instead curve elements of a curve over this field.